### ISSUE TITLE
Optional support for new xdg background portal

### ISSUE THREAD

**Issue by AsavarTzeth** [wwmm/easyeffects#535]
Thanks to a new portal interface ([org.freedesktop.impl.portal.Background](https://github.com/flatpak/xdg-desktop-portal/blob/master/data/org.freedesktop.impl.portal.Background.xml)) it should now be possible to solve this old issue: https://github.com/flathub/com.github.wwmm.pulseeffects/issues/18

In time this should ideally be used everywhere, whenever possible. It does however add a dependency on `xdg-desktop-portal >= 1.4.0`, which means distributions providing older versions is the primary reason it can't always be relied upon at this time.

Would it be possible to check for the availability of this DBus API at runtime and if available use that instead of writing `pulseeffects-service.desktop` directly?

I am no expert but if you need any more information or assistance I will try my best to provide it.

**Comment by wwmm** [comment]
At least at compile time it is possible to check if it is available as they use pkg-config. I have no idea about runtime. And I also have no idea about how to use it. I think I will need to see a code example. The dbus service file we just have to put in the right place. DBUS and GTK take care of the rest on their own.

**Comment by wwmm** [comment]
Wait... The DBUS service file is not related to the autostart. It only starts PE background service when the user clicks on PE lunch icon. The autostart switch in PE menu does not use the DBUS service file. It creates the autostart file directly as shown here https://github.com/wwmm/pulseeffects/blob/8e95621d6e11a0376b04e12beed7402e942c5db9/src/general_settings_ui.cpp#L140

**Comment by wwmm** [comment]
Forget the previsou messages. I was thinking about another service file XD hahaha

**Comment by wwmm** [comment]
These are the places where changes have to be made https://github.com/wwmm/pulseeffects/blob/8e95621d6e11a0376b04e12beed7402e942c5db9/src/general_settings_ui.cpp#L140 and https://github.com/wwmm/pulseeffects/blob/8e95621d6e11a0376b04e12beed7402e942c5db9/src/general_settings_ui.cpp#L124. When we find out how to use this portal it is only a matter of using a  preprocessor `#ifdef` based on the availability of `xdg-desktop-portal >= 1.4.0` at compile time. I imagine that the flatpak package is the only one that needs to use this portal. So we could have a build time option that you could use to make meson compile PE with this portal.

**Comment by AsavarTzeth** [comment]
Sorry for confusing things. Just to clarify xdg-desktop-portal is a host dependency, not a bundled depedency. It is a dependency of Flatpak itself as you can see with `pacman -Qi xdg-desktop-portal`.

I will try to find some example though.

**Comment by AsavarTzeth** [comment]
Surely a compile time check would be worthless since we build it once an re-use it it on all distributions, which again might not have `xdg-desktop-portal >= 1.4.0`.

**Comment by wwmm** [comment]
Maybe we are not even supposed to compile PE with it. Maybe its pkg-config file is meant to be used when flatpak is compiled. I don't know.

Assuming everything can be done through DBUS we could use GLIB wrappers for DBUS as it is already done for realtimekit. When we find a code example things will be more clear.

**Comment by AsavarTzeth** [comment]
Yes, that is what I failed to convey. If we take Arch Linux for example you are meant to install this as a regular package, before you can install flatpak apps.

I should have mentioned that these were DBus API:s from the beginning. Back to searching for that example then.

**Comment by AsavarTzeth** [comment]
Not sure if it is enough but I found this in old mailing list post. Unfortunately the example is not for this particular portal. I thought I would post this just in case.

> Yeah, the thing to use here is the background portal (which is kinda
new, so beware it may not be everywhere yet).
"org.freedesktop.impl.portal.Background" is actually the backend (thus
the ".impl"), so what you want to use is
org.freedesktop.portal.Background.

> The docs for that are here:
  https://github.com/flatpak/xdg-desktop-portal/blob/master/data/org.freedesktop.portal.Background.xml

> The portal apis are a bit complex since they are long-running and
abortable. But what it boils down to is calling the
RequestBackground() method with autostart=true and
commandline=/app/bin/data-collect-app (and possibly
dbus-activatable=true if the collector is a dbus service, which may be
a good way to talk to it). This will ask the user if it is ok for the
app to run in the background, and if so record this for the future (to
allow the app to run in the background)) and generate an autostart
file with the right contents to have your app started on login.
Once this is done you will get a signal on the request object that
this succeeded (or was denied).

> Some portal are automatic behind other apis, but this one you need to
call manually. Here is some portal calling code you can use as a
template:
https://gitlab.gnome.org/GNOME/glib/blob/master/gio/gopenuriportal.c#L244

> There has been discussions of creating a wrapper library for these
APIs, but there is none atm.

*Substitute __data-collect-app__ with __pulseeffects__*

**Comment by AsavarTzeth** [comment]
I'm hoping we can resurrect this old issue and perhaps finally close https://github.com/flathub/com.github.wwmm.pulseeffects/issues/18.

Since this was first posted things have developed. We now have [`libportal`](https://github.com/flatpak/libportal), which should greatly simplify the use of the background portal.

> When we find out how to use this portal it is only a matter of using a preprocessor #ifdef

We can still do this, now with `libportal`. For comparison that was done with [`eog`](https://gitlab.gnome.org/GNOME/eog/-/commit/ca73abc36d3704c216e213642b4c5b370a55a279), following [issue 137](https://gitlab.gnome.org/GNOME/eog/-/issues/137). Even though that issue seems based on fundamental misunderstandings, I can think of reasons for it for keeping it optional.

That said, I do think it should be enabled by default, when the dependency is found. Portals do work outside of flatpaks and sometimes bring the advantage of desktop integrations and a more uniform user experience. If that leads to issues, you can always change it later on.

Unfortunately I still can't find any app that uses the background portal specifically. That combined with my lacking programming skills and current mental capacity, resulted in failure when I tried to implement this myself. It should be a simple enough job when you know what you're doing.

Other apps that uses this can be found in the "Required by" section at [packages.archlinux.org](https://archlinux.org/packages/extra/x86_64/libportal/). Documentation ships with the library in the form of gtk-doc generated html.

I will be available for testing and anything else you might need me for.

**Comment by wwmm** [comment]
I have been a little busy but I will try to remember to take a look at this in the next days.

**Comment by wwmm** [comment]
I tried to follow what is done in this link https://github.com/flatpak/xdg-desktop-portal/blob/669e91c2774850944ac8f11881de4bd69831c182/tests/background.c#L67 but it does not even compile:
```
/usr/include/libportal/portal-helpers.h:57:19: error: expected unqualified-id before ‘export’
   57 |   XdpParentExport export;
```

I don't know what I am missing...

**Comment by wwmm** [comment]
Just including the header is enough to have a failed compilation. Even with `dependency('libportal')` in the meson file...

**Comment by AsavarTzeth** [comment]
Then it would seem you came as far as I did. I too never figured out that error and I just assumed it was due to my own inexperience. At least, now I don't feel as stupid :|

I will see if I can get some answers and get back to you.

**Comment by vchernin** [comment]
Ok, I'd like to try to get this working because it is still useful for EasyEffects in Flatpak.

One detail which is somewhat new: apparently for the most part libportal/xdg desktop portals shouldn't be used for non-sandboxed apps https://github.com/flatpak/libportal/issues/33. So libportal should probably be an optional dependency for now.

I am really not sure how exactly an implementation should work, I am not very good at programming nor am I very familar with how portals work. But I will try.

Do you remember what you tried to do here? https://github.com/wwmm/easyeffects/issues/535#issuecomment-779288409. Maybe I can try to redo that now.

**Comment by wwmm** [comment]
>One detail which is somewhat new: apparently for the most part libportal/xdg desktop portals shouldn't be used for non-sandboxed apps flatpak/libportal#33. So libportal should probably be an optional dependency for now.

Unless they decide to support non sandboxed apps it will have to be an optional dependency forever.

>Do you remember what you tried to do here? #535 (comment). Maybe I can try to redo that now.

I couldn't even try to do something. Just including the library header was enough to get a compilation error.

**Comment by vchernin** [comment]
> Unless they decide to support non sandboxed apps it will have to be an optional dependency forever.

Yeah, I do think eventually that will change, but who knows...

> I couldn't even try to do something. Just including the library header was enough to get a compilation error. 

Do you remember what file needed to header? I'd probably need to include libportal in the right meson.build file too right? I can try to build it locally and see what happens.

**Comment by wwmm** [comment]
>Do you remember what file needed the header? 

Probably this one https://github.com/wwmm/easyeffects/blob/master/include/general_settings_ui.hpp as this is the class handling the autostart switch.

**Comment by vchernin** [comment]
Ok, I seem to be able to get it to build with the background header. https://github.com/wwmm/easyeffects/pull/1115

I'll try to actually get it to do something useful later.

**Comment by vchernin** [comment]
So for now the Flatpak will be patched to use libportal in the nest Flathub release. This is mostly due to https://github.com/flatpak/xdg-desktop-portal/issues/579. Because of that portal issue the background portal doesn't work properly on the host, so it means two codepaths would need to be maintained which is sub optimal.

See https://github.com/wwmm/easyeffects/pull/1141 for more discussion.

**Comment by wwmm** [comment]
The flatpak build has been using libportal for a while. We can close this issue.
